File: wrappers/java/src/main/java/org/hyperledger/indy/sdk/pool/Pool.java
package org.hyperledger.indy.sdk.pool;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import org.hyperledger.indy.sdk.IndyException;
import org.hyperledger.indy.sdk.IndyJava;
import org.hyperledger.indy.sdk.LibIndy;
import org.hyperledger.indy.sdk.ParamGuard;

import com.sun.jna.Callback;

/**
 * pool.rs API
 */

/**
 * High level wrapper around SDK Pool functionality.
 */
public class Pool extends IndyJava.API implements AutoCloseable {

	private final int poolHandle;

	private Pool(int poolHandle) {

		this.poolHandle = poolHandle;
	}

	/**
	 * Gets the handle for the pool instance.
	 * 
	 * @return The handle for the pool instance.
	 */
	public int getPoolHandle() {

		return this.poolHandle;
	}

	/*
	 * STATIC CALLBACKS
	 */

	/**
	 * Callback used when openPoolLedger completes.
	 */
	private static Callback openPoolLedgerCb = new Callback() {

		@SuppressWarnings({"unused", "unchecked"})
		public void callback(int xcommand_handle, int err, int pool_handle) {

			CompletableFuture<Pool> future = (CompletableFuture<Pool>) removeFuture(xcommand_handle);
			if (! checkResult(future, err)) return;

			Pool pool = new Pool(pool_handle);

			Pool result = pool;
			future.complete(result);
		}
	};

	/**
	 * Callback used when void function completes.
	 */
	private static Callback voidCb = new Callback() {

		@SuppressWarnings({"unused", "unchecked"})
		public void callback(int xcommand_handle, int err) {

			CompletableFuture<Void> future = (CompletableFuture<Void>) removeFuture(xcommand_handle);
			if (! checkResult(future, err)) return;

			Void result = null;
			future.complete(result);
		}
	};

	/**
	 * Callback used when listPools completes.
	 */
	private static Callback listPoolsCb = new Callback() {

		@SuppressWarnings({"unused", "unchecked"})
		public void callback(int xcommand_handle, int err, String metadata) {

			CompletableFuture<String> future = (CompletableFuture<String>) removeFuture(xcommand_handle);
			if (! checkResult(future, err)) return;

			String result = metadata;
			future.complete(result);
		}
	};
	
	/*
	 * STATIC METHODS
	 */

	/**
	 * Creates a new local pool ledger configuration that can be used later to connect pool nodes.
	 * 
	 * @param configName Name of the pool ledger configuration.
	 * @param config Pool configuration json. if NULL, then default config will be used.
	 * @return A future that does not resolve a value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<Void> createPoolLedgerConfig(
			String configName,
			String config) throws IndyException {

		ParamGuard.notNullOrWhiteSpace(configName, "configName");		
		
		CompletableFuture<Void> future = new CompletableFuture<Void>();
		int commandHandle = addFuture(future);

		int result = LibIndy.api.indy_create_pool_ledger_config(
				commandHandle, 
				configName, 
				config,
				voidCb);

		checkResult(future, result);

		return future;
	}

	/**
	 * Opens pool ledger and performs connecting to pool nodes.
	 * 
	 * @param configName Name of the pool ledger configuration.
	 * @param config Runtime pool configuration json. Optional. If NULL, then default config will be used.
	 *               Example:
	 * {
	 *     "timeout": int (optional), timeout for network request (in sec).
	 *     "extended_timeout": int (optional), extended timeout for network request (in sec).
	 *     "preordered_nodes": array(string) -  (optional), names of nodes which will have a priority during request sending:
	 *          ["name_of_1st_prior_node",  "name_of_2nd_prior_node", .... ]
	 *          This can be useful if a user prefers querying specific nodes.
	 *          Assume that `Node1` and `Node2` nodes reply faster.
	 *          If you pass them Libindy always sends a read request to these nodes first and only then (if not enough) to others.
	 *          Note: Nodes not specified will be placed randomly.
	 *     "number_read_nodes": int (optional) - the number of nodes to send read requests (2 by default)
	 *          By default Libindy sends a read requests to 2 nodes in the pool.
	 *          If response isn't received or `state proof` is invalid Libindy sends the request again but to 2 (`number_read_nodes`) * 2 = 4 nodes and so far until completion.
	 * }
	 *
	 * @return A future that resolves to an opened Pool instance.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<Pool> openPoolLedger(
			String configName,
			String config) throws IndyException {

		ParamGuard.notNullOrWhiteSpace(configName, "configName");	
		
		CompletableFuture<Pool> future = new CompletableFuture<Pool>();
		int commandHandle = addFuture(future);

		int result = LibIndy.api.indy_open_pool_ledger(
				commandHandle, 
				configName, 
				config, 
				openPoolLedgerCb);

		checkResult(future, result);

		return future;
	}

	/**
	 * Refreshes a local copy of a pool ledger and updates pool nodes connections.
	 * 
	 * @param pool The pool to refresh.
	 * @return A future that does not resolve a value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	private static CompletableFuture<Void> refreshPoolLedger(
			Pool pool) throws IndyException {

		ParamGuard.notNull(pool, "pool");	
		
		CompletableFuture<Void> future = new CompletableFuture<Void>();
		int commandHandle = addFuture(future);

		int handle = pool.getPoolHandle();

		int result = LibIndy.api.indy_refresh_pool_ledger(
				commandHandle, 
				handle,
				voidCb);

		checkResult(future, result);

		return future;
	}

	/**
	 * Closes opened pool ledger, opened nodes connections and frees allocated resources.
	 * 
	 * @param pool The pool to close.
	 * @return A future that does not resolve a value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	private static CompletableFuture<Void> closePoolLedger(
			Pool pool) throws IndyException {

		ParamGuard.notNull(pool, "pool");	
		
		CompletableFuture<Void> future = new CompletableFuture<Void>();
		int commandHandle = addFuture(future);

		int handle = pool.getPoolHandle();

		int result = LibIndy.api.indy_close_pool_ledger(
				commandHandle, 
				handle,
				voidCb);

		checkResult(future, result);

		return future;
	}

	/**
	 * Deletes created pool ledger configuration.
	 * 
	 * @param configName Name of the pool ledger configuration to delete.
	 * @return A future that does not resolve a value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<Void> deletePoolLedgerConfig(
			String configName) throws IndyException {

		ParamGuard.notNullOrWhiteSpace(configName, "configName");	
		
		CompletableFuture<Void> future = new CompletableFuture<Void>();
		int commandHandle = addFuture(future);

		int result = LibIndy.api.indy_delete_pool_ledger_config(
				commandHandle, 
				configName,
				voidCb);

		checkResult(future, result);

		return future;
	}

	/**
	 * Set PROTOCOL_VERSION to specific version.
	 *
	 * There is a global property PROTOCOL_VERSION that used in every request to the pool and
	 * specified version of Indy Node which Libindy works.
	 *
	 * By default PROTOCOL_VERSION=1.
	 *
	 * @param protocolVersion Protocol version will be used:
	 *      1 - for Indy Node 1.3
	 *      2 - for Indy Node 1.4 and greater
	 *
	 * @return A future that does not resolve a value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<Void> setProtocolVersion(
			int protocolVersion) throws IndyException {

		CompletableFuture<Void> future = new CompletableFuture<Void>();
		int commandHandle = addFuture(future);

		int result = LibIndy.api.indy_set_protocol_version(
				commandHandle,
				protocolVersion,
				voidCb);

		checkResult(future, result);

		return future;
	}

	/**
	 * Lists names of created pool ledgers
	 *
	 * @return A future resolving to a list of pools: [{
	 *     "pool": string - Name of pool ledger stored in the wallet.
	 *   }]
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<String> listPools() throws IndyException {

		CompletableFuture<String> future = new CompletableFuture<String>();
		int commandHandle = addFuture(future);

		int result = LibIndy.api.indy_list_pools(
				commandHandle,
				listPoolsCb);

		checkResult(future, result);

		return future;
	}

	/*
	 * INSTANCE METHODS
	 */

	/**
	 * Refreshes a local copy of a pool ledger and updates pool nodes connections.
	 * 
	 * @return A future that does not resolve a value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public CompletableFuture<Void> refreshPoolLedger(
			) throws IndyException {

		return refreshPoolLedger(this);
	}

	/**
	 * Closes opened pool ledger, opened nodes connections and frees allocated resources.
	 * 
	 * @return A future that does not resolve a value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public CompletableFuture<Void> closePoolLedger(
			) throws IndyException {

		return closePoolLedger(this);
	}

	@Override
	public void close() throws InterruptedException, ExecutionException, IndyException {
		closePoolLedger().get();
	}
}

File: libindy/src/api/pool.rs

use indy_api_types::{ErrorCode, CommandHandle, PoolHandle, INVALID_POOL_HANDLE};
use crate::commands::{Command, CommandExecutor};
use crate::commands::pool::PoolCommand;
use crate::domain::pool::{PoolConfig, PoolOpenConfig};
use indy_api_types::errors::prelude::*;
use indy_utils::ctypes;
use indy_api_types::validation::Validatable;

use serde_json;
use libc::c_char;

/// Creates a new local pool ledger configuration that can be used later to connect pool nodes.
///
/// #Params
/// config_name: Name of the pool ledger configuration.
/// config (optional): Pool configuration json. if NULL, then default config will be used. Example:
/// {
///     "genesis_txn": string (optional), A path to genesis transaction file. If NULL, then a default one will be used.
///                    If file doesn't exists default one will be created.
/// }
///
/// #Returns
/// Error code
///
/// #Errors
/// Common*
/// Ledger*
#[no_mangle]
pub extern fn indy_create_pool_ledger_config(command_handle: CommandHandle,
                                             config_name: *const c_char,
                                             config: *const c_char,
                                             cb: Option<extern fn(command_handle_: CommandHandle,
                                                                  err: ErrorCode)>) -> ErrorCode {
    trace!("indy_create_pool_ledger_config: >>> config_name: {:?}, config: {:?}", config_name, config);

    check_useful_c_str!(config_name, ErrorCode::CommonInvalidParam2);
    check_useful_opt_json!(config, ErrorCode::CommonInvalidParam3, PoolConfig);
    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam4);

    trace!("indy_create_pool_ledger_config: entities >>> config_name: {:?}, config: {:?}", config_name, config);

    let result = CommandExecutor::instance()
        .send(Command::Pool(PoolCommand::Create(
            config_name,
            config,
            Box::new(move |result| {
                let err = prepare_result!(result);
                trace!("indy_create_pool_ledger_config:");
                cb(command_handle, err)
            })
        )));

    let res = prepare_result!(result);

    trace!("indy_create_pool_ledger_config: <<< res: {:?}", res);

    res
}

/// Opens pool ledger and performs connecting to pool nodes.
///
/// Pool ledger configuration with corresponded name must be previously created
/// with indy_create_pool_ledger_config method.
/// It is impossible to open pool with the same name more than once.
///
/// config_name: Name of the pool ledger configuration.
/// config (optional): Runtime pool configuration json.
///                         if NULL, then default config will be used. Example:
/// {
///     "timeout": int (optional), timeout for network request (in sec).
///     "extended_timeout": int (optional), extended timeout for network request (in sec).
///     "preordered_nodes": array<string> -  (optional), names of nodes which will have a priority during request sending:
///         ["name_of_1st_prior_node",  "name_of_2nd_prior_node", .... ]
///         This can be useful if a user prefers querying specific nodes.
///         Assume that `Node1` and `Node2` nodes reply faster.
///         If you pass them Libindy always sends a read request to these nodes first and only then (if not enough) to others.
///         Note: Nodes not specified will be placed randomly.
///     "number_read_nodes": int (optional) - the number of nodes to send read requests (2 by default)
///         By default Libindy sends a read requests to 2 nodes in the pool.
///         If response isn't received or `state proof` is invalid Libindy sends the request again but to 2 (`number_read_nodes`) * 2 = 4 nodes and so far until completion.
///     "socks_proxy": string (optional) - ZMQ socks proxy host name and port (example: proxy1.intranet.company.com:1080)
/// }
///
/// #Returns
/// Handle to opened pool to use in methods that require pool connection.
///
/// #Errors
/// Common*
/// Ledger*
#[no_mangle]
pub extern fn indy_open_pool_ledger(command_handle: CommandHandle,
                                    config_name: *const c_char,
                                    config: *const c_char,
                                    cb: Option<extern fn(command_handle_: CommandHandle,
                                                         err: ErrorCode,
                                                         pool_handle: PoolHandle)>) -> ErrorCode {
    trace!("indy_open_pool_ledger: >>> config_name: {:?}, config: {:?}", config_name, config);

    check_useful_c_str!(config_name, ErrorCode::CommonInvalidParam2);
    check_useful_opt_validatable_json!(config, ErrorCode::CommonInvalidParam3, PoolOpenConfig);
    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam4);

    trace!("indy_open_pool_ledger: entities >>> config_name: {:?}, config: {:?}", config_name, config);

    let result = CommandExecutor::instance()
        .send(Command::Pool(PoolCommand::Open(
            config_name,
            config,
            Box::new(move |result| {
                let (err, pool_handle) = prepare_result_1!(result, INVALID_POOL_HANDLE);
                trace!("indy_open_pool_ledger: pool_handle: {:?}", pool_handle);
                cb(command_handle, err, pool_handle)
            })
        )));

    let res = prepare_result!(result);

    trace!("indy_open_pool_ledger: <<< res: {:?}", res);

    res
}

/// Refreshes a local copy of a pool ledger and updates pool nodes connections.
///
/// #Params
/// handle: pool handle returned by indy_open_pool_ledger
///
/// #Returns
/// Error code
///
/// #Errors
/// Common*
/// Ledger*
#[no_mangle]
pub extern fn indy_refresh_pool_ledger(command_handle: CommandHandle,
                                       handle: PoolHandle,
                                       cb: Option<extern fn(command_handle_: CommandHandle,
                                                            err: ErrorCode)>) -> ErrorCode {
    trace!("indy_refresh_pool_ledger: >>> handle: {:?}", handle);

    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam3);

    trace!("indy_refresh_pool_ledger: entities >>> handle: {:?}", handle);

    let result = CommandExecutor::instance()
        .send(Command::Pool(PoolCommand::Refresh(
            handle,
            Box::new(move |result| {
                let err = prepare_result!(result);
                trace!("indy_refresh_pool_ledger:");
                cb(command_handle, err)
            })
        )));

    let res = prepare_result!(result);

    trace!("indy_refresh_pool_ledger: <<< res: {:?}", res);

    res
}

/// Lists names of created pool ledgers
///
/// #Params
///
/// #Returns
/// Error code
///
/// #Errors
#[no_mangle]
pub extern fn indy_list_pools(command_handle: CommandHandle,
                              cb: Option<extern fn(command_handle_: CommandHandle,
                                                   err: ErrorCode,
                                                   pools: *const c_char)>) -> ErrorCode {
    trace!("indy_list_pools: >>>");

    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam2);

    trace!("indy_list_pools: entities >>>");

    let result = CommandExecutor::instance()
        .send(Command::Pool(PoolCommand::List(boxed_callback_string!("indy_list_pools", cb, command_handle))));

    let res = prepare_result!(result);

    trace!("indy_list_pools: <<< res: {:?}", res);

    res
}

/// Closes opened pool ledger, opened nodes connections and frees allocated resources.
///
/// #Params
/// handle: pool handle returned by indy_open_pool_ledger.
///
/// #Returns
/// Error code
///
/// #Errors
/// Common*
/// Ledger*
#[no_mangle]
pub extern fn indy_close_pool_ledger(command_handle: CommandHandle,
                                     handle: PoolHandle,
                                     cb: Option<extern fn(command_handle_: CommandHandle,
                                                          err: ErrorCode)>) -> ErrorCode {
    trace!("indy_close_pool_ledger: >>> handle: {:?}", handle);

    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam3);

    trace!("indy_close_pool_ledger: entities >>> handle: {:?}", handle);

    let result = CommandExecutor::instance()
        .send(Command::Pool(PoolCommand::Close(
            handle,
            Box::new(move |result| {
                let err = prepare_result!(result);
                trace!("indy_close_pool_ledger:");
                cb(command_handle, err)
            })
        )));

    let res = prepare_result!(result);

    trace!("indy_close_pool_ledger: <<< res: {:?}", res);

    res
}

/// Deletes created pool ledger configuration.
///
/// #Params
/// config_name: Name of the pool ledger configuration to delete.
///
/// #Returns
/// Error code
///
/// #Errors
/// Common*
/// Ledger*
#[no_mangle]
pub extern fn indy_delete_pool_ledger_config(command_handle: CommandHandle,
                                             config_name: *const c_char,
                                             cb: Option<extern fn(command_handle_: CommandHandle,
                                                                  err: ErrorCode)>) -> ErrorCode {
    trace!("indy_delete_pool_ledger_config: >>> config_name: {:?}", config_name);

    check_useful_c_str!(config_name, ErrorCode::CommonInvalidParam2);
    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam3);

    trace!("indy_delete_pool_ledger_config: entities >>> config_name: {:?}", config_name);

    let result = CommandExecutor::instance()
        .send(Command::Pool(PoolCommand::Delete(
            config_name,
            Box::new(move |result| {
                let err = prepare_result!(result);
                trace!("indy_delete_pool_ledger_config:");
                cb(command_handle, err)
            })
        )));

    let res = prepare_result!(result);

    trace!("indy_delete_pool_ledger_config: <<< res: {:?}", res);

    res
}

/// Set PROTOCOL_VERSION to specific version.
///
/// There is a global property PROTOCOL_VERSION that used in every request to the pool and
/// specified version of Indy Node which Libindy works.
///
/// By default PROTOCOL_VERSION=1.
///
/// #Params
/// protocol_version: Protocol version will be used:
///     1 - for Indy Node 1.3
///     2 - for Indy Node 1.4 and greater
///
/// #Returns
/// Error code
///
/// #Errors
/// Common*
#[no_mangle]
pub extern fn indy_set_protocol_version(command_handle: CommandHandle,
                                        protocol_version: usize,
                                        cb: Option<extern fn(command_handle_: CommandHandle,
                                                             err: ErrorCode)>) -> ErrorCode {
    trace!("indy_set_protocol_version: >>> protocol_version: {:?}", protocol_version);

    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam3);

    trace!("indy_set_protocol_version: entities >>> protocol_version: {:?}", protocol_version);

    let result = CommandExecutor::instance()
        .send(Command::Pool(
            PoolCommand::SetProtocolVersion(
            protocol_version,
            Box::new(move |result| {
                let err = prepare_result!(result);
                trace!("indy_set_protocol_version:");
                cb(command_handle, err)
            })
        )));

    let res = prepare_result!(result);

    trace!("indy_set_protocol_version: <<< res: {:?}", res);

    res
}
